code/datums/components/fullauto.dm

```diff
@@ -0,0 +1,284 @@
+		source.click_intercept_time = world.time
+		return
+
+	if(source.mob.in_throw_mode)
+		return
+	if(!isturf(source.mob.loc)) //No firing inside lockers and stuff.
+		return
+	if(get_dist(source.mob, _target) < 2) //Adjacent clicking.
+		return
+
+	if(isnull(location)) //Clicking on a screen object.
+		if(_target.plane != CLICKCATCHER_PLANE) //The clickcatcher is a special case. We want the click to trigger then, under it.
+			return //If we click and drag on our worn backpack, for example, we want it to open instead.
+		_target = params2turf(modifiers["screen-loc"], get_turf(source.eye), source)
+		if(!_target)
+			CRASH("Failed to get the turf under clickcatcher")
+
+	if(SEND_SIGNAL(src, COMSIG_AUTOFIRE_ONMOUSEDOWN, source, _target, location, control, params) & COMPONENT_AUTOFIRE_ONMOUSEDOWN_BYPASS)
+		return
+
+	source.click_intercept_time = world.time //From this point onwards Click() will no longer be triggered.
+
+	if(autofire_stat & (AUTOFIRE_STAT_IDLE))
+		CRASH("on_mouse_down() called with [autofire_stat] autofire_stat")
+	if(autofire_stat & AUTOFIRE_STAT_FIRING)
+		stop_autofiring() //This can happen if we click and hold and then alt+tab, printscreen or other such action. MouseUp won't be called then and it will keep autofiring.
+
+	target = _target
+	target_loc = get_turf(target)
+	mouse_parameters = params
+	start_autofiring()
+
+
+//Dakka-dakka
+/datum/component/automatic_fire/proc/start_autofiring()
+	if(autofire_stat == AUTOFIRE_STAT_FIRING)
+		return //Already pew-pewing.
+	autofire_stat = AUTOFIRE_STAT_FIRING
+
+	if(auto_delay_timer) //This shouldn't be happening, so let's stack_trace it and remove it if nothing is caught.
+		stack_trace("start_autofiring called with a non-null auto_delay_timer")
+		if(!deltimer(auto_delay_timer))
+			addtimer(CALLBACK(src, .proc/keep_trying_to_delete_timer, auto_delay_timer), 1)
+		auto_delay_timer = null
+
+	clicker.mouse_override_icon = 'icons/effects/mouse_pointers/supplypod_target.dmi'
+	clicker.mouse_pointer_icon = clicker.mouse_override_icon
+
+	if(mouse_status == AUTOFIRE_MOUSEUP) //See mouse_status definition for the reason for this.
+		RegisterSignal(clicker, COMSIG_CLIENT_MOUSEUP, .proc/on_mouse_up)
+		mouse_status = AUTOFIRE_MOUSEDOWN
+
+	RegisterSignal(shooter, COMSIG_MOB_SWAP_HANDS, .proc/stop_autofiring)
+
+	if(isgun(parent))
+		var/obj/item/gun/shoota = parent
+		if(!shoota.on_autofire_start(shooter)) //This is needed because the minigun has a do_after before firing and signals are async.
+			stop_autofiring()
+			return
+	if(autofire_stat != AUTOFIRE_STAT_FIRING)
+		return //Things may have changed while on_autofire_start() was being processed, due to do_after's sleep.
+
+	if(!process_shot()) //First shot is processed instantly.
+		return //If it fails, such as when the gun is empty, then there's no need to schedule a second shot.
+	auto_delay_timer = addtimer(CALLBACK(src, .proc/process_shot), autofire_shot_delay, TIMER_STOPPABLE|TIMER_LOOP)
```

Comment:
This needs to be handled by a subsystem instead of by a looping timer.
There's a TGMC dev already working on it if you don't want to do it yourself.